presto-native-execution/presto_cpp/main/PeriodicTaskManager.cpp

```diff
@@ -420,6 +437,83 @@ void PeriodicTaskManager::addCacheStatsUpdateTask() {
       "cache_counters");
 }
 
+// Example Apple shrink function. Shrinks if free memory is less than 2GiB
+#ifdef __APPLE__
+bool exampleAppleShrinkFunc() {
+  vm_size_t page_size;
+  mach_port_t mach_port;
+  mach_msg_type_number_t count;
+  vm_statistics64_data_t vm_stats;
+  mach_port = mach_host_self();
+  unsigned long long free_memory = 0;
+
+  count = sizeof(vm_stats) / sizeof(natural_t);
+  if (KERN_SUCCESS == host_page_size(mach_port, &page_size) &&
+      KERN_SUCCESS ==
+          host_statistics64(
+              mach_port, HOST_VM_INFO, (host_info64_t)&vm_stats, &count)) {
+    free_memory = (int64_t)vm_stats.free_count * (int64_t)page_size;
+  }
+  return free_memory ? free_memory < (2 << 30) : false;
+}
+#endif
+
+// Example Linux docker container shrink function.
+// Shrinks if free memory is less than 5 GiB
+#ifdef __linux__
+bool exampleLinuxShrinkFunc() {
+  static long long containerMemoryLimit = 0;
+  struct sysinfo memInfo;
+  long long inUseMemory;
+  unsigned long long minReserveGB = 5;
+
+  sysinfo(&memInfo);
+
+  if (!containerMemoryLimit) {
+    auto containerMemoryLimitFile =
+        folly::File("/sys/fs/cgroup/memory/memory.limit_in_bytes", O_RDONLY);
+    // Enough storage for the container memory limit
+    std::array<char, 50> buf;
+    if (!folly::readNoInt(
+            containerMemoryLimitFile.fd(), buf.data(), buf.size())) {
+      LOG(INFO)
+          << "/sys/fs/cgroup/memory/memory.limit_in_bytes not found or is empty. Using system limit";
+      containerMemoryLimit = memInfo.totalram;
+    } else if (sscanf(buf.data(), "%" SCNu64, &containerMemoryLimit) != 1) {
+      LOG(INFO) << "No integer value read. Using system limit";
+      containerMemoryLimit = memInfo.totalram;
+    }
+    LOG(INFO) << "System memory limit: " << containerMemoryLimit;
+  }
+  inUseMemory = (memInfo.totalram - memInfo.freeram) * memInfo.mem_unit;
+  return (containerMemoryLimit - inUseMemory) < (minReserveGB << 30);
+}
+#endif
+
+void PeriodicTaskManager::addCacheShrinkTask() {
+#if defined(__APPLE__)
+  addTask(
```

Comment:
If we set this to 0 to disable (and not have the default) - should we skip creating the task?
If it is 0 currently then it looks like it will run without wait.

I think we should have some way to disable it.

Reply:
Sure

Reply:
Done